File: credit/credit.c
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void checkCard(string id);
string checkCardType(string id);
int validateCard(string id);

int main(void)
{
    long cardID = get_long("Number: ");
    string id = malloc(sizeof(long int) + 1);
    sprintf(id, "%ld", cardID);

    checkCard(id);
}

void checkCard(string id)
{
    if (strlen(id) < 13)
    {
        printf("INVALID\n");
        return;
    }

    if (validateCard(id) == 0)
    {
        printf("INVALID\n");
    }
    else
    {
        string cardType = checkCardType(id);
        printf("%s\n", cardType);
    }
}

string checkCardType(string id)
{
    int length = strlen(id);
    if (length == 15)
    {
        if (id[0] == '3' && (id[1] == '4' || id[1] == '7'))
        {
            return "AMEX";
        }
    }
    else if (length == 16)
    {
        if (id[0] == '5' &&
            (id[1] == '1' || id[1] == '2' || id[1] == '3' || id[1] == '4' || id[1] == '5'))
        {
            return "MASTERCARD";
        }
        else if (id[0] == '4')
        {
            return "VISA";
        }
    }
    else if (length == 13 && id[0] == '4')
    {
        return "VISA";
    }

    return "INVALID";
}

int validateCard(string id)
{
    int sum1 = 0;
    int sustracted = 0;
    for (int i = strlen(id) - 1; i > -1; i--)
    {
        if (sustracted % 2 == 0)
        {
            sum1 = sum1 + id[i] - '0';
        }
        else
        {
            int temp = (id[i] - '0') * 2;
            if (temp < 10)
            {
                sum1 = sum1 + temp;
            }
            else
            {
                sum1 = sum1 + (temp % 10) + (temp / 10);
            }
        }
        sustracted++;
    }

    // printf("sum1 = %i\n", sustracted);

    if (sum1 % 10 == 0)
    {
        return 1;
    }

    return 0;
}

File: sentimental-credit/credit.py
from cs50 import get_int


def main():
    cardID = str(get_int("Number: "))

    checkCard(cardID)


def checkCard(cardID):
    if (len(cardID) < 13):
        print("INVALID")

    if (validateCard(cardID) == 0):
        print("INVALID")
    else:
        cardType = checkCardType(cardID)
        print(cardType)


def validateCard(cardID):
    sum1 = 0
    substracted = 0
    for i in range((len(cardID)-1), -1, -1):
        if (substracted % 2 == 0):
            sum1 = sum1 + int(cardID[i])
        else:
            temp = int(cardID[i]) * 2
            if temp < 10:
                sum1 = sum1 + temp
            else:
                sum1 = sum1 + (temp % 10) + (temp // 10)
        substracted += 1

    if sum1 % 10 == 0:
        print("validateCard pass")
        return 1

    return 0


def checkCardType(cardID):
    length = len(cardID)
    if length == 15:
        if cardID[0] == '3' and (cardID[1] == '4' or cardID[1] == '7'):
            return "AMEX"
    elif length == 16:
        if cardID[0] == '5' and (cardID[1] in ['1', '2', '3', '4', '5']):
            return "MASTERCARD"
        elif cardID[0] == '4':
            return "VISA"
    elif length == 13 and cardID[0] == '4':
        return "VISA"

    return "INVALID"


main()
